matrix_sdk_base/src/store/sled_store/mod.rs

```diff
@@ -550,6 +648,129 @@ impl SledStore {
+                            let first_event = timeline.events.first().unwrap().event_id();
+                            let last_event = timeline.events.last().unwrap().event_id();
+
+                            let found_first = if let Some((batch_idx, current_position)) =
+                                event_id_to_position
+                                    .get((room.as_str(), first_event.as_str()).encode())?
+                                    .map(|e| decode_batch_idx_position(&e).unwrap())
+                            {
+                                Some((false, true, batch_idx, current_position))
+                            } else {
+                                None
+                            };
+
+                            let found_last = if let Some((batch_idx, current_position)) =
+                                event_id_to_position
+                                    .get((room.as_str(), last_event.as_str()).encode())?
+                                    .map(|e| decode_batch_idx_position(&e).unwrap())
+                            {
+                                let current_position = current_position - timeline.events.len() + 1;
+
+                                Some((true, false, batch_idx, current_position))
+                            } else {
+                                None
+                            };
+
+                            // The store already knowns all events of this slice.
+                            if found_first.is_some() && found_last.is_some() {
+                                continue;
+                            }
+
+                            found_first.or(found_last)
+                        };
+
+                        // Lookup the previous or next batch
+                        let expandable = if let Some(expandable) = expandable {
+                            Some(expandable)
+                        } else if let Some((batch_idx, position)) = end_token_to_batch_idx_position
+                            .remove((room.as_str(), timeline.start.as_str()).encode())?
+                            .map(|e| decode_batch_idx_position(&e).unwrap())
+                        {
+                            Some((false, true, batch_idx, position))
+                        } else if let Some(end) = &timeline.end {
+                            if let Some((batch_idx, position)) = start_token_to_batch_idx_position
+                                .remove((room.as_str(), end.as_str()).encode())?
+                                .map(|e| decode_batch_idx_position(&e).unwrap())
+                            {
+                                let position = position - timeline.events.len();
+                                Some((true, false, batch_idx, position))
+                            } else {
+                                None
+                            }
+                        } else {
+                            None
+                        };
```

Comment:
This works (without checking all four possible combinations) because of storing all events "backwards" right? So you check if the `start` token is in the end `Tree` or if the `end` token is in the start `Tree`?